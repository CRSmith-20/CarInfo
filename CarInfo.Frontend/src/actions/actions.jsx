import axios from 'axios';
import React from 'react';

const GET_AVAILABLE_MAKES = process.env.REACT_APP_API_BASE_URL + '/make'
const GET_MODELS_FOR_MAKES = process.env.REACT_APP_API_BASE_URL + '/models/'
const GET_YEARS_FOR_MODEL = process.env.REACT_APP_API_BASE_URL + '/years/'
const GET_CAR_DETAILS = process.env.REACT_APP_API_BASE_URL + '/details/'

export async function getAvailableMakes() {
    var data = await axios.get(GET_AVAILABLE_MAKES).then(
        response => {
            return response.data;
        }
    )
    .catch(error => "API Request failed with: " + error);

    
    return data;
}

export async function getModelsForMake(maker) {
    var data = await axios.get(GET_MODELS_FOR_MAKES + maker).then(
        response => {
            return response.data
        }
    );

    return data;
}

export async function getYearsForModel(model) {
    var data = await axios.get(GET_YEARS_FOR_MODEL + model).then(
        response => {
            let yearsWithId = []
            response.data.map(function(pair) {
                yearsWithId.push({ID: pair["ID"], "Year": pair["Year"]});
            })
            return yearsWithId
        }
    );

    return data;
}

export async function getCarDetails(id) {
    var data = await axios.get(GET_CAR_DETAILS + id).then(
        response => {
            let carInfo = {}
            let engineData = []

            response.data.map(function(details) {
                if(Object.keys(carInfo).length == 0){
                    carInfo = {Drive: details["Drive"], Transmission: details["Transmission"]}
                }

                engineData.push({Style: details["EngineStyle"], Horsepower: details["Horsepower"], Rpm: details["EngineRpm"], CityMPG: details["CityMpg"], HighwayMPG: details["HighwayMpg"]})
            })
            return {CarInfo: carInfo, EngineData: engineData}
        }
    );
    return data;
}

export default getAvailableMakes;