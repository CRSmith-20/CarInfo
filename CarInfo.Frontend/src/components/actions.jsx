import axios from 'axios';
import React from 'react';

const GET_AVAILABLE_MAKES = process.env.REACT_APP_API_BASE_URL + '/make'
const GET_MODELS_FOR_MAKES = process.env.REACT_APP_API_BASE_URL + '/model/'

export async function getAvailableMakes() {
    var data = await axios.get(GET_AVAILABLE_MAKES).then(
        response => {
            console.log(response);
            return { makes: response.data };
        }
    );
    
    return data;
}

export async function getModelsForMake(maker) {
    var data = await axios.get(GET_MODELS_FOR_MAKES + maker).then(
        response => {
            return { models: response.data}
        }
    );

    return data;
}