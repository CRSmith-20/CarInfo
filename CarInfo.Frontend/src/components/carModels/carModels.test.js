import {shallow, mount} from 'enzyme';
import CarModels from './carModels';
import axios from "axios";
import { BrowserRouter as Router } from 'react-router-dom';
import { getModelsForMake } from '../../actions/actions';
import { act } from "react-dom/test-utils";
import toJson from 'enzyme-to-json';
import { createMemoryHistory } from 'history'

jest.mock("../../actions/actions", () => ({getModelsForMake: jest.fn()}));

let models;
let match;
let wrapper;
let history;

beforeEach(() => {
    models =  [
        "Sonata", "Elantra", "Hotel", "Trivago"
    ]
    match = { params: { make: 'Horrible' } }

    history = createMemoryHistory();
    history.location.key = 'consistent'; //needed for snapshot as it's unique per run

});

describe('<CarModels />', () => {
    describe('when component initalizes', () => {
        it('should initalize to loading', () => {
            const test = shallow(<Router><CarModels /></Router>);
            expect(toJson(test)).toMatchSnapshot();
        });
    });
    
    describe('when component loads', () => {
        it('should render models with links', async () => {
            
            await act(async () => {
                getModelsForMake.mockResolvedValueOnce(models);
                wrapper = mount(
                    <Router>
                        <CarModels history={history} match={match} />
                    </Router>
                );
            });
    
            wrapper.update();

            expect(toJson(wrapper)).toMatchSnapshot();
            expect(wrapper.find('Link').length).toBe(4);
        })
    });
    
    describe('when the api call fails', () => {
        it('should render a ErrorDisplay', async () => {                      
            await act(async () => {
                getModelsForMake.mockImplementationOnce(() => Promise.resolve(["Error", new Error("Generic Failure")]));
                wrapper = mount(<Router><CarModels history={history} match={match} /></Router>);
            });
    
            wrapper.update();

            expect(toJson(wrapper)).toMatchSnapshot();
            expect(wrapper.find('ErrorDisplay').length).toBe(1);
        })
    });
})