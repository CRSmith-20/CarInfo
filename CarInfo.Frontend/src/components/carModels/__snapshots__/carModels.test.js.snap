// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CarModels /> when component initalizes should initalize to loading 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <CarModels />
</Router>
`;

exports[`<CarModels /> when component loads should render models with links 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <CarModels
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "consistent",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "consistent",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
      match={
        Object {
          "params": Object {
            "make": "Horrible",
          },
        }
      }
    >
      <div>
        <div>
          <div
            key="Sonata"
          >
            <Link
              to="/model/Sonata"
            >
              <LinkAnchor
                href="/model/Sonata"
                navigate={[Function]}
              >
                <a
                  href="/model/Sonata"
                  onClick={[Function]}
                >
                  Sonata
                </a>
              </LinkAnchor>
            </Link>
          </div>
          <div
            key="Elantra"
          >
            <Link
              to="/model/Elantra"
            >
              <LinkAnchor
                href="/model/Elantra"
                navigate={[Function]}
              >
                <a
                  href="/model/Elantra"
                  onClick={[Function]}
                >
                  Elantra
                </a>
              </LinkAnchor>
            </Link>
          </div>
          <div
            key="Hotel"
          >
            <Link
              to="/model/Hotel"
            >
              <LinkAnchor
                href="/model/Hotel"
                navigate={[Function]}
              >
                <a
                  href="/model/Hotel"
                  onClick={[Function]}
                >
                  Hotel
                </a>
              </LinkAnchor>
            </Link>
          </div>
          <div
            key="Trivago"
          >
            <Link
              to="/model/Trivago"
            >
              <LinkAnchor
                href="/model/Trivago"
                navigate={[Function]}
              >
                <a
                  href="/model/Trivago"
                  onClick={[Function]}
                >
                  Trivago
                </a>
              </LinkAnchor>
            </Link>
          </div>
        </div>
        <button
          onClick={[Function]}
        >
          Return to Makes
        </button>
      </div>
    </CarModels>
  </Router>
</BrowserRouter>
`;

exports[`<CarModels /> when the api call fails should render a ErrorDisplay 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <CarModels
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "consistent",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "consistent",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
      match={
        Object {
          "params": Object {
            "make": "Horrible",
          },
        }
      }
    >
      <ErrorDisplay
        title="Select Model"
      >
        <div>
          <h1
            style={
              Object {
                "textAlign": "center",
              }
            }
          >
            Select Model
          </h1>
          <Alert
            closeLabel="Close alert"
            key="error"
            show={true}
            style={
              Object {
                "textAlign": "center",
              }
            }
            transition={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": Object {
                  "appear": false,
                  "in": false,
                  "mountOnEnter": false,
                  "timeout": 300,
                  "unmountOnExit": false,
                },
                "render": [Function],
              }
            }
            variant="warning"
          >
            <Fade
              appear={false}
              in={true}
              mountOnEnter={false}
              style={
                Object {
                  "textAlign": "center",
                }
              }
              timeout={300}
              unmountOnExit={true}
            >
              <Transition
                addEndListener={[Function]}
                appear={false}
                enter={true}
                exit={true}
                in={true}
                mountOnEnter={false}
                onEnter={[Function]}
                onEntered={[Function]}
                onEntering={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                onExiting={[Function]}
                style={
                  Object {
                    "textAlign": "center",
                  }
                }
                timeout={300}
                unmountOnExit={true}
              >
                <div
                  className="fade alert alert-warning show"
                  role="alert"
                  style={
                    Object {
                      "textAlign": "center",
                    }
                  }
                >
                  An error has occurred while loading, please refresh and try again.
                </div>
              </Transition>
            </Fade>
          </Alert>
        </div>
      </ErrorDisplay>
    </CarModels>
  </Router>
</BrowserRouter>
`;
